### PR TITLE
LPS-196390 Add a popup when changing tab without saving

### PR DESCRIPTION
**Pull request by ces-luannguyen**
Hi @locpham97 @tototrinh @ces-quanhuynh ,
Please help me review this ticket.

### Motivation
[LPS-196390 ](https://liferay.atlassian.net/browse/LPS-196390)Add a popup when changing tab without saving

### Solution
LPS-196390
I bind an event to navigation bar. This event will fire a popup if there's changes in permission modal.

Thanks,
Luan Nguyen

### CODE REVIEW

File: modules/apps/portlet-configuration/portlet-configuration-web/src/main/resources/META-INF/resources/edit_permissions.jsp [hunk 1]
```diff
@@ -30,9 +30,16 @@ PortletConfigurationPermissionPropagation portletConfigurationPermissionPropagat
 
 <div class="cadmin edit-permissions portlet-configuration-edit-permissions">
 	<div class="portlet-configuration-body-content">
-		<clay:navigation-bar
-			navigationItems="<%= portletConfigurationPermissionsDisplayContext.getNavigationItems() %>"
-		/>
+		<div>
+			<react:component
+				module="js/FolderDocumentPermissionTab"
```
Comment (locpham97) [inline modules/apps/portlet-configuration/portlet-configuration-web/src/main/resources/META-INF/resources/edit_permissions.jsp:35]: I think it should be PropagationPermissionBar or PermissionNavigationBar or something like that

File: modules/apps/portlet-configuration/portlet-configuration-web/src/main/resources/META-INF/resources/js/FolderDocumentPermissionTab.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	navigationItems: _navigationItems,
+	portletNamespace: _portletNamespace
+}) {
+
+	const alertMessage = document.getElementById(
+		_portletNamespace+'alertMessage'
+	);
+	const saveButton = document.getElementById(
+		_portletNamespace+'saveButton');
+	function onChangeTab(event, href) {
```
Comment (locpham97) [inline modules/apps/portlet-configuration/portlet-configuration-web/src/main/resources/META-INF/resources/js/FolderDocumentPermissionTab.js:15]: should have a space in here

File: modules/apps/portlet-configuration/portlet-configuration-web/src/main/resources/META-INF/resources/js/FolderDocumentPermissionTab.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+import ClayLink from '@clayui/link';
+import ClayNavigationBar from '@clayui/navigation-bar';
+import React from 'react';
+export default function FolderDocumentPermissionTab({
+	navigationItems: _navigationItems,
+	portletNamespace: _portletNamespace
+}) {
+
+	const alertMessage = document.getElementById(
+		_portletNamespace+'alertMessage'
```
Comment (locpham97) [inline modules/apps/portlet-configuration/portlet-configuration-web/src/main/resources/META-INF/resources/js/FolderDocumentPermissionTab.js:11]: _portletNameSpace + 'alertMessage'

File: modules/apps/portlet-configuration/portlet-configuration-web/src/main/resources/META-INF/resources/js/FolderDocumentPermissionTab.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+						window.location.href = href;
+					},
+				},
+			],
+			status: 'warning',
+			title: Liferay.Language.get('discard-changes')+'?',
+		});
+	}
+	return (
+
```
Comment (locpham97) [inline modules/apps/portlet-configuration/portlet-configuration-web/src/main/resources/META-INF/resources/js/FolderDocumentPermissionTab.js:53]: remove space

**Comment by ces-luannguyen** [comment]
@locpham97 @tototrinh I updated code, please help double check.

Thanks,
Luan Nguyen

File: modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties [hunk 5]
```diff
@@ -14826,6 +14826,7 @@ saq-configuration-service-metrics=Metrics
 saq-configuration-service-signature=Service Signature
 saturday-abbreviation=S
 save=Save
+changing-tab-without-save-helper=Changing tab without save is an action impossible to revert. All changes will be lost and it won't be possible to recover it. Are you sure you want to discard those changes?
```
Comment (locpham97) [inline modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties]: place it right order before you start buildlang

File: modules/apps/portlet-configuration/portlet-configuration-web/src/main/resources/META-INF/resources/js/PermissionNavigationBar.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+	const saveButton = document.getElementById(
+		_portletNamespace + 'saveButton'
+	);
+	function onChangeTab(event, href) {
+		if (!alertMessage.classList.contains('hide')) {
+			openPopUp(href);
+		}
+		else {
+			window.location.href = href;
+		}
```
Comment (tototrinh) [inline modules/apps/portlet-configuration/portlet-configuration-web/src/main/resources/META-INF/resources/js/PermissionNavigationBar.js:26]: The popup will not be shown when editing the permissions of a folder with the propagation setting set to false. Please double-check. Thanks.

**Comment by ces-luannguyen** [comment]
Hi @locpham97 @tototrinh @ces-quanhuynh 
I updated code. Please help me review it.

Thanks,
Luan Nguyen

File: modules/apps/portlet-configuration/portlet-configuration-web/src/main/resources/META-INF/resources/edit_permissions.jsp [hunk 7]
```diff
@@ -63,6 +65,10 @@ PortletConfigurationPermissionPropagation portletConfigurationPermissionPropagat
 		</c:if>
 
 		<aui:form action="<%= portletConfigurationPermissionsDisplayContext.getUpdateRolePermissionsURL() %>" cssClass="container-fluid container-fluid-max-xl" method="post" name="fm">
+			<aui:input name="count" type="hidden">
```
Comment (locpham97) [inline modules/apps/portlet-configuration/portlet-configuration-web/src/main/resources/META-INF/resources/edit_permissions.jsp:68]: <aui:input name="count" type="hidden" />

File: modules/apps/portlet-configuration/portlet-configuration-web/src/main/resources/META-INF/resources/edit_permissions.jsp [hunk 8]
```diff
@@ -323,4 +329,66 @@ PortletConfigurationPermissionPropagation portletConfigurationPermissionPropagat
...
+
+				if (
+					!target.classList.contains('active') &&
+					Number(count.value) !== 0
+				) {
+					openPopUp(target.href);
+				}
+				else {
+					window.location.href = target.href;
+				}
```
Comment (tototrinh) [inline modules/apps/portlet-configuration/portlet-configuration-web/src/main/resources/META-INF/resources/edit_permissions.jsp:390]: I think we can move `event.preventDefault()` to `if` block instead. There is no need to add an `else` block.

**Comment by ces-locpham** [comment]
Hi @ces-luannguyen  , 
I tested your solution with localStorage( SessionStorage). It works fine, you can reference my solution:
1de2b833d60020beacae27227938d69b9408f456 and replace it by SessionStorage

**Comment by ces-quanhuynh** [comment]
@tototrinh @locpham97 
This new PR https://github.com/tototrinh/liferay-portal/pull/81 will continue handle this PR. 
Please review the new one.
Thanks.
cc @ces-luannguyen

**Pull request closed without merging**